docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+
+
+## API evolution
+
+The main changes:
+
+* we deprecate createExternalLoadBalancer, and replace it with visibility
+* we move publicIPs to the Status
+* we add a field hostPort to each service port
+ 
+Mappings:
+
+`visibility` is mappable:
+
+```
+new.visibility = old.createExternalLoadBalancer ? "loadbalancer" : "cluster"
+```
+
+```
+old.createExternalLoadBalancer = (new.visibility == 'loadbalancer')
+```
+
+`publicIPs` will be copied to the Spec if set on the Status, and from the Spec to the Status if not set on the Status.
+Once each cloud provider sets the Status, we will stop copying from the Spec to the Status.
+
+For v1 final:
+* we should remove publicIPs from the Spec, and require that this information is retrieved through the Status.
+However, we would need to change the openstack code to auto-allocate a VIP.
+* we should remove createExternalLoadBalancer.
+
+## kube-proxy pseudocode
+
+```
+for each service {
+    for each service.port {
+        // listen on cluster
+        listen on <internalMagicIp>:0
+        proxyPort = figureOutTheRandomPortWeJustGotAssigned()
+        openPortal(service.portalIP, port.port, proxyPort)
+        
+
+        if service.spec.visibility == "loadbalancer" || (service.spec.visibility == "" && service.createExternalLoadBalancer=true) {
+          if (port.hostPort != 0) {
+            openLoadBalancerPortals(service, port.port, port.hostPort) // uses cloud-provider to figure out correct iptables
+          }
+        } else if service.spec.visibility == "public" {
+          if (port.hostPort != 0) {
+            openPublicPortal(<publicMagicIp>, port.port, port.hostPort)
+          }
+        } else if service.spec.visibility == "" {
+          for each service.publicIP {  // deprecated
+            openPublicPortal(publicIP, port.port, proxyPort) // uses cloud-provider to decide srcip vs dstip (gclb vs rax)
+          }
+        }
+   }
+}
+```
+
+
+## kube-proxy iptables
+
+As part of the implementation of kube-proxy's internal magic, kube-proxy allocates ports that are used locally by
+iptables.  These ports are not visible other than to kube-proxy & iptables.  We need to be sure that these ports will
+not conflict with the cluster-wide ports.  To do that, we must make sure that kube-proxy differentiates between
+internal & external traffic, and uses these internal ports only with internal traffic.
+
+There are a few possible implementation strategies here:
+
+1. Use `sysctl -w net.ipv4.conf.all.route_localnet=1`, and then bind kube-proxy to 127.0.0.1 for internal traffic.  Requires kernel >= 3.6
+1. Use addrtype with iptables (external: `! --src-type local --dst-type local`, internal: `--src-type local --dst-type local`)
+1. Use the network interface to identify traffic (internal: lo, cbr0; external: eth0)
+1. Use the container CIDR range to identify internal (container) traffic (e.g. 10.244.0.0/16).  Requires fixing that CIDR.
```

Comment:
> On Apr 29, 2015, at 9:12 PM, Quinton Hoole <notifications@github.com> wrote:
> 
> In docs/load_balancer.md:
> 
> > +```
> > +
> > +
> > +## kube-proxy iptables
> > +
> > +As part of the implementation of kube-proxy's internal magic, kube-proxy allocates ports that are used locally by
> > +iptables.  These ports are not visible other than to kube-proxy & iptables.  We need to be sure that these ports will
> > +not conflict with the cluster-wide ports.  To do that, we must make sure that kube-proxy differentiates between
> > +internal & external traffic, and uses these internal ports only with internal traffic.
> > +
> > +There are a few possible implementation strategies here:
> > +
> > +1. Use `sysctl -w net.ipv4.conf.all.route_localnet=1`, and then bind kube-proxy to 127.0.0.1 for internal traffic.  Requires kernel >= 3.6
> > +1. Use addrtype with iptables (external: `! --src-type local --dst-type local`, internal: `--src-type local --dst-type local`)
> > +1. Use the network interface to identify traffic (internal: lo, cbr0; external: eth0)
> > +1. Use the container CIDR range to identify internal (container) traffic (e.g. 10.244.0.0/16).  Requires fixing that CIDR.
> On Wed, Apr 29, 2015 at 6:03 PM, Alex Robinson <notifications@github.com> wrote: In docs/load_balancer.md <https://github.com/GoogleCloudPlatform/kubernetes/pull/7447#discussion_r29396774> : > +``` > + > + > +## kube-proxy iptables > + > +As part of the implementation of kube-proxy's internal magic, kube-proxy allocates ports that are used locally by > +iptables. These ports are not visible other than to kube-proxy & iptables. We need to be sure that these ports will > +not conflict with the cluster-wide ports. To do that, we must make sure that kube-proxy differentiates between > +internal & external traffic, and uses these internal ports only with internal traffic. > + > +There are a few possible implementation strategies here: > + > +1. Use `sysctl -w net.ipv4.conf.all.route_localnet=1`, and then bind kube-proxy to 127.0.0.1 for internal traffic. Requires kernel >= 3.6 > +1. Use addrtype with iptables (external: `! --src-type local --dst-type local`, internal: `--src-type local --dst-type local`) > +1. Use the network interface to identify traffic (internal: lo, cbr0; external: eth0) > +1. Use the container CIDR range to identify internal (container) traffic (e.g. 10.244.0.0/16). Requires fixing that CIDR. This option may be not be viable forever -- some folks are pushing to allocate IPs dynamically rather than giving each node a static CIDR.
> I don't think that dynamic pod IP allocation works with non-SDN network environments? i.e. typically a real network admin has to configure real routers with static routes? (caveat - I'm not totally familiar with exactly how the routing of our Pod-IP's works - Abi assures me that they're real routed packets).

You can ask DHCP for an IP, at which point a specifically advanced local network can do whatever it wants.  I know a few folks who are looking at that path, and it's not quite SDN but not quite static allocation.

> …
> —
> Reply to this email directly or view it on GitHub.
>